src/veil/proofofstake/stakeinput/stakeinput.cpp

```diff
@@ -0,0 +1,114 @@
+#include <chainparams.h>
+#include "stakeinput.h"
+
+std::string StakeInputTypeToString(StakeInputType t)
+{
+    if (t == STAKE_ZEROCOIN) {
+        return "ZerocoinStake";
+    } else if (t == STAKE_RINGCT) {
+        return "RingCtStake";
+    } else if (t == STAKE_RINGCTCANDIDATE) {
+        return "RingCtStakeCandidate";
+    }
+    return "error-type";
+}
+
+int HeightToModifierHeight(int nHeight)
+{
+    //Nearest multiple of KernelModulus that is over KernelModulus bocks deep in the chain
+    return (nHeight - Params().KernelModulus()) - (nHeight % Params().KernelModulus()) ;
+}
+
+// Use the PoW hash or the PoS hash
+uint256 GetHashFromIndex(const CBlockIndex* pindexSample)
+{
+    if (pindexSample->IsProofOfWork())
+        return pindexSample->GetBlockPoWHash();
+
+    uint256 hashProof = pindexSample->GetBlockPoSHash();
+    return hashProof;
+}
+
+// As the sampling gets further back into the chain, use more bits of entropy. This prevents the ability to significantly
+// impact the modifier if you create one of the most recent blocks used. For example, the contribution from the first sample
+// (which is 101 blocks from the coin tip) will only be 2 bits, thus only able to modify the end result of the modifier
+// 4 different ways. As the sampling gets further back into the chain (take previous sample height - nSampleCount*6 for the height
+// it will be) it has more entropy that is assigned to that specific sampling. The further back in the chain, the less
+// ability there is to have any influence at all over the modifier calculations going forward from that specific block.
+// The bits taper down as it gets deeper into the chain so that the precomputability is less significant.
+int GetSampleBits(int nSampleCount)
+{
+    switch(nSampleCount) {
+        case 0:
+            return 2;
+        case 1:
+            return 4;
+        case 2:
+            return 8;
+        case 3:
+            return 16;
+        case 4:
+            return 32;
+        case 5:
+            return 64;
+        case 6:
+            return 128;
+        case 7:
+            return 64;
+        case 8:
+            return 32;
+        case 9:
+            return 16;
+        default:
+            return 0;
+    }
+}
```

Comment:
An interesting variation to PoS 2.0 entropy. I do understand that this is moved code by the way, just hadn't come across this yet.